UPGRADING.md

```diff
@@ -0,0 +1,101 @@
+Upgrading
+=========
+
+### From 1.x to 2.x
+
+#### Name Change
+
+`InteropResolver` was renamed to `ContainerResolver` - the method-signature has not changed, so you only need
+to update your imports to reference the new name.
+
+#### MiddlewareInterface Removed
+
+The built-in `MiddlewareInterface` has been removed, and you need to select one of the two PSR-15 interfaces,
+as described below.
+
+If you used `callable` middleware, you should expect errors/exceptions, as middleman retains support for
+`callable`, but now requires a PSR-15 compatible method-signature, as described below.
+
+#### PSR-15 Middleware
+
+`mindplay/middleman^2` adopts the [PSR-15](https://github.com/http-interop/http-middleware) `0.2` interfaces.
+
+This changes the middleware signature substantially, from the legacy signature:
+
+    function (RequestInterface $request, ResponseInterface $response, callable $next): ResponseInterface
+
+To the PSR-15 signature:
+
+    function (RequestInterface|ServerRequestInterface $request, DelegateInterface $delegate): ResponseInterface
+
+PSR-15 introduces two distinct interfaces: `MiddlewareInterface` for processing `RequestInterface`, and
+`ServerMiddlewareInterface` for processing `ServerRequestInterface`.
+
+Because the Response object no longer passes down through the middleware stack, you will need to port your
+middleware components from the old signature to one of the new signatures.
+
+For example, if you had something like the following:
+
+```php
+use mindplay\middleman\MiddlewareInterface;
+
+class MyMiddleware implements MiddlewareInterface
+{
+    public function __invoke(RequestInterface $request, ResponseInterface $response, $next)
+    {
+        if ($request instanceof ServerRequestInterface) {
+            if ($request->getUri()->getPath() === "/test") {
+                return $response->withBody(...);
+            }
+        }
+
+        return $next($request, $response->withHeader("X-Foo", "bar");
+    }
+}
+```
+
+You would need to change several things:
+
+```php
+use Interop\Http\Middleware\ServerMiddlewareInterface;
+
+class MyMiddleware implements ServerMiddlewareInterface
+{
+    public function process(ServerRequestInterface $request, DelegateInterface $next)
+    {
+        if ($request->getUri()->getPath() === "/test") {
+            return new Response()->withBody(...);
+        }
+
+        $response = $next->process($request);
+
+        if (! $response->hasHeader("X-Foo")) {
+            $response = $response->withHeader("X-Foo", "bar");
+        }
+
+        return $next->process($response);
```

Comment:
How can the "process" method take both a ServerRequestInterface, and a ResponseInterface?
```
$response = $next->process($request);
return $next->process($response);
```